### PR TITLE
Draft logAndthrow API

### PR DESCRIPTION
**Pull request by hemanttanwar**

### CODE REVIEW

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 1]
```diff
@@ -159,6 +159,22 @@ public void log(String format, Object... args) {
...
 
+    /**
+     * This will log, if error log level is enabled , and throw the runtime exception.
+     * @param runtimeException to be thrown. It will do nothing if {@code null} is provided.
+     */
+    public void logAndThrow(RuntimeException runtimeException) {
+        if(runtimeException == null) {
+            return;
+        }
+        // By default we are treating runtime exception as error ?
```
Comment (hemanttanwar) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: By default log level in ClientLogger is DISABLED_LEVEL.  Here we are saying that for logAndthrow we want to change loglevel to "ERROR_LEVEL" because we want to print it if it is turned on via ConfigurationManager.getConfiguration()

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 2]
```diff
@@ -159,6 +159,22 @@ public void log(String format, Object... args) {
...
+        level = ERROR_LEVEL;
+        if (canLogAtLevel(level)) {
```
Comment (hemanttanwar) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: By default log level in ClientLogger is DISABLED_LEVEL.  Here we are saying that for logAndthrow we want to change loglevel to "ERROR_LEVEL" because we want to print it if it is turned on via ConfigurationManager.getConfiguration()

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 3]
```diff
@@ -155,21 +155,36 @@ public ClientLogger asError() {
...
      * @param format Format-able message.
      * @param args Arguments for the message, if an exception is being logged last argument is the throwable.
      */
-    private void performLogging(String format, Object... args) {
+    private void performLogging(String format, boolean removeThroawable, Object... args) {
         // If the logging level is less granular than verbose remove the potential throwable from the args.
-        if (configurationLevel > VERBOSE_LEVEL) {
+        if (removeThroawable && configurationLevel > VERBOSE_LEVEL) {
             args = attemptToRemoveThrowable(args);
         }
```
Comment (hemanttanwar) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: We do want to log Runtime Exception  if someone has called logAndThrow().   But we want to remove Runtime Exception if some one is calling regular log() ….  Thus we need Boolean flag, removeThroawable,  for this flipping

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 4]
```diff
@@ -155,21 +155,47 @@ public ClientLogger asError() {
...
+            level = ERROR_LEVEL;
+        }
+
+        // it will only log if log level is enabled in configuration
+        if (canLogAtLevel(level)) {
+            performLogging(runtimeException.getClass().getName(), false, runtimeException);
+        }
+        
+        // Change level back to what it was.
+        level = currentLevel;
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: I'm not convinced we need this extra complexity. Could we just call `logger.error("", runtimeException)` without changing level or modifying any of the other code?

Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: I did had question myself on this, It is good suggestion to use logger.error.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 5]
```diff
@@ -159,6 +159,23 @@ public void log(String format, Object... args) {
...
+     * @throws RuntimeException which is requested by this call.
+     */
+    public void logAndThrow(RuntimeException runtimeException) {
+        if (runtimeException == null) {
+            return;
+        }
+
+        // it will only log if log level is enabled in configuration
+        if (canLogAtLevel(level)) {
+            logger.error("", runtimeException);
```
Comment (srnagar) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: Can we use `runtimeException.getMessage()` instead of empty string?

File: core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java [hunk 6]
```diff
@@ -58,6 +59,7 @@
  */
 public class Poller<T> {
 
+    ClientLogger logger = new ClientLogger(Poller.class);
```
Comment (srnagar) [inline core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java]:
```suggestion
    private final ClientLogger logger = new ClientLogger(Poller.class);
```

**Comment by JonathanGiles** [comment]
Hemant, please take into consideration the feedback that @srnagar has provided here. Also, before this is merged I want @alzimmermsft to weigh in with his thoughts.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 7]
```diff
@@ -159,6 +159,23 @@ public void log(String format, Object... args) {
         }
     }
 
+    /**
+     * This will log {@link RuntimeException}, if logging is enabled , and throw the runtime exception.
+     * @param runtimeException to be thrown. It will do nothing if {@code null} is provided.
+     * @throws RuntimeException which is requested by this call.
+     */
+    public void logAndThrow(RuntimeException runtimeException) {
```
Comment (alzimmermsft) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: I'm not sure exactly how we would want to surface the API but should there be a logAndThrowError, maybe just throwError, and logAndThrowWarning? The logging spec states that error should only be used in cases where the application won't be able to recover, such as an out of memory exception, where as warning should be used in recoverable cases, such as a REST call failing.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 8]
```diff
@@ -159,6 +159,23 @@ public void log(String format, Object... args) {
...
+        if (runtimeException == null) {
+            return;
+        }
+
+        // it will only log if log level is enabled in configuration
+        if (canLogAtLevel(level)) {
```
Comment (alzimmermsft) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: This check won't always work as expected and won't comply with the changes made in PR #4194.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java (hunk 5 above)
Comment (alzimmermsft) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: This is needed as the exception will be redacted from logging if the logging level doesn't include debugging/verbose logs.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: The reasoning behind this is, we are going to use this to throw unchecked Exception. If we make it Throwable and making it checked exception , it will cause caller to handle it or throw it. 
For example
logger.logAndThrow(new IllegalArgumentException("Null value for poll operation is not allowed."));

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java (hunk 5 above)
Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: In case where client has not enabled logging. We honour that and do not expose anything in logs. We show more bias towads not looging PII.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JonathanGiles) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: RuntimeException is fine for now.

Comment (JonathanGiles) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: Let's wait until there is a need for this API.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
+(canLogAtLevel(level)// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: Revert this change

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java (hunk 8 above)
Comment (JonathanGiles) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: @hemanttanwar Did you reconcile what this comment means? It would be good to see a follow-up comment to clarify we're on the same page.

Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: Spoke to @alzimmermsft and he is fine with the changes done to handle this case.

File: core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java [hunk 10]
```diff
@@ -159,6 +159,23 @@ public void log(String format, Object... args) {
         }
     }
 
+    /**
+     * This will log {@link RuntimeException}, if logging is enabled , and throw the runtime exception.
```
Comment (conniey) [inline core/azure-core/src/main/java/com/azure/core/util/logging/ClientLogger.java]: nit: `enabled , and` to `enabled, and`
